File: algorithmic-toolbox/src/week2/java/CoveringSegments.java
package week2.java;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

class CoveringSegments {

    public static void main(final String[] args) {
        try (final Scanner in = new Scanner(System.in)) {
            final int[][] ns = new int[in.nextInt()][];
            for (int i = 0; i < ns.length; i++) {
                ns[i] = new int[] { in.nextInt(), in.nextInt() };
            }

            Arrays.sort(ns, (a, b) -> (a[1] != b[1]) ? a[1] - b[1] : b[0] - a[0]);
            final List<Integer> result = new ArrayList<>();

            int max = ns[0][1];
            result.add(max);
            for (int i = 1; i < ns.length; i++) {
                if (max < ns[i][0]) {
                    max = ns[i][1];
                    result.add(max);
                }
            }

            System.out.println(result.size());
            for (int i = 0; i < result.size(); i++) {
                if (i > 0) {
                    System.out.print(" ");
                }

                System.out.print(result.get(i));
            }
        }
    }

}

File: algorithmic-toolbox/src/week2/scala/CoveringSegments.scala
object CoveringSegments {
  def main(args: Array[String]) = {
    val in = new java.util.Scanner(System.in)
    val ps = getPoints(
      (1 to in.nextInt)
        .map(n => (in.nextInt, in.nextInt))
        .sortWith((a, b) => if (a._2 == b._2) (a._1 > b._1) else (a._2 < b._2)))
    println(s"${ps.size}\n${ps.mkString(" ")}")
  }

  def getPoints(ps: Seq[(Int, Int)]) = {
    def getPoints(rem: Seq[(Int, Int)], max: Int, acc: List[Int]): List[Int] = {
      if (rem.isEmpty) acc
      else if (max < rem.head._1) getPoints(rem.tail, rem.head._2, acc :+ rem.head._2)
      else getPoints(rem.tail, max, acc)
    }

    getPoints(ps.tail, ps.head._2, List(ps.head._2))
  }
}